20160218 overcommit plugin

sosiouxme opened the pull request:
The first commit is the main commentary re https://trello.com/c/rRN7m0zv/546-5-override-container-resource-request-based-on-limit

The second commit I tacked on because this links to LimitRange objects and those are defined in limits.html which really IMHO belongs in the admin guide. This should address at least part of https://github.com/openshift/openshift-docs/issues/809 - nevertheless I give separate commits in order to make it easy to hold this back for further consideration if desired.

sosiouxme commented:
@abhgupta for your review

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+Note that these overrides have no effect if no limits have
+been set on containers. link:./limits.html[Create a LimitRange
+object] with default limits (per individual project, or in the
+link:./selfprovisioned_projects.html#template-for-new-projects[project
```
New home at:

https://docs.openshift.org/latest/admin_guide/managing_projects.html#modifying-the-template-for-new-projects

adellape commented on dev_guide/compute_resources.adoc:
```diff
@@ -137,3 +137,11 @@ Containers:
+
+== Project Resource Limits
+
+link:../admin_guide/limits.html-limits[Resource limits can be
```
s/-/#/

We also have to consider the openshift-dedicated and openshift-online distros, which don't include the admin_guide and install_config, therefore breaking these links in those builds. We've generally used ifdefs for this sort of situation.

adellape commented:
@sosiouxme Some comments inline.

+1 on the moving of the "Resource Limits" topic.

Found a few (probably new, may need a rebase here) spots that still need xrefs updated for the new `admin_guide` path:

`dev_guide/deployments.adoc` (consider the Online/Dedicated builds that can't link to `admin_guide`, though)
`install_config/configuring_admission_control.adoc`

Side note: Once this merges, I'll need to follow up with Harrison about a redirect (for the topic location change) on the Origin site, and then again later for the downstream sites once it's cherry-picked to the respective branches.

adellape commented:
FYI, @bfallonf's PR here is about to update the current `dev_guide/limits.adoc`:

https://github.com/openshift/openshift-docs/pull/1704

adellape commented:
@sosiouxme Needs a rebase.

sosiouxme commented:
Thanks for the bump, I'll bump it up my TODOs a bit...

On Thu, Apr 14, 2016 at 10:58 AM, Alex Dellapenta <notifications@github.com>
wrote:

> @sosiouxme <https://github.com/sosiouxme> Needs a rebase.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/openshift-docs/pull/1612#issuecomment-209986659>
>

sosiouxme commented:
@adellape I think this is ready now.

sosiouxme commented:
@adellape PTAL

adellape commented on dev_guide/compute_resources.adoc:
```diff
@@ -279,3 +279,13 @@ To specify compute resources via the CLI:
 ----
 ====
+
+ifdef::openshift-enterprise,openshift-origin[]
```
@sosiouxme You can add `openshift-dedicated` here, we have a Cluster Administrator guide for Dedicated now:

https://docs.openshift.com/dedicated/latest/admin_guide/index.html

adellape commented on dev_guide/deployments.adoc:
```diff
@@ -599,9 +599,16 @@ items is required:
+- A limit range defined in your project, where the
+defaults from the `*LimitRange*` object apply to pods created during the
+deployment process.
+endif::[]
```
Same re: `openshift-dedicated` on these.

adellape commented on _build_cfg.yml:
```diff
@@ -388,7 +388,7 @@ Topics:
   - Name: Setting Quotas
     File: quota
-  - Name: Setting Limit Ranges
+  - Name: Resource Limits
```
"Setting Limit Ranges" was the new name given to this topic when it was moved via https://github.com/openshift/openshift-docs/pull/2059/files#diff-e612d75a53fc180c5c4512806a7b1c31.

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+scheduled based on the 1Gi request being available on the node, but could use
+up to 2Gi; so it is 200% overcommitted.
+
+If OpenShift administrators would like to control the level of overcommit
```
s/OpenShift/{product-title}/

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+LimitRange] specifying limits and defaults, this adjusts the container
+limit and request to achieve the desired level of overcommit.
+
+This requires configuring the ClusterResourceOverride admission controller in the
```
So far we've marked up the other admission controllers like:

```
`*ClusterResourceOverride*`
```

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+limit and request to achieve the desired level of overcommit.
+
+This requires configuring the ClusterResourceOverride admission controller in the
+master-config.yaml as in the following example (reuse the existing configuration tree
```
```
*_master-config.yaml_*
```

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+
+This requires configuring the ClusterResourceOverride admission controller in the
+master-config.yaml as in the following example (reuse the existing configuration tree
+if it exists, or introduce absent elements as needed).
```
s/./:/

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+          cpuRequestToLimitPercent: 25     <3>
+          limitCPUToMemoryPercent: 200     <4>
+----
+<1> This is the plugin name; case matters and anything but an exact match for a plugin name is ignored.
```
s/plugin/plug-in/

x2

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+<4> (optional, positive integer) If a container memory limit has been specified or defaulted, the CPU limit is overridden to a percentage of the memory limit, with a 100 percentage scaling 1Gi of RAM to equal 1 CPU core. This is processed prior to overriding CPU request (if configured).
+====
+
+After changing the master config, a master restart is required.
```
s/config/configuration/

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+link:./managing_projects.html#modifying-the-template-for-new-projects[project
+template]) in order to ensure that the overrides apply.
+
+Note also that after overrides, the container limits/requests must still
```
limits and requests

adellape commented on admin_guide/overcommit.adoc:
```diff
@@ -127,6 +127,68 @@ exist.
+to allow infrastructure components to be configured independently of
+overrides) by editing the project and adding the following annotation:
+
+    quota.openshift.io/cluster-resource-override-enabled: "false"
```
```
----
quota.openshift.io/cluster-resource-override-enabled: "false"
----
```

adellape commented:
@sosiouxme Nits inline.

Also, should we add this to the list of tasks described in the Admissions Controller arch topic?

https://docs.openshift.com/enterprise/latest/architecture/additional_concepts/admission_controllers.html

Also also: This was OSE 3.2 only?

sosiouxme commented:
@adellape Thanks, addressed those issues. I added it to the Admission Controllers page, not sure we want to go out of our way to point people to this solution but it *is* an admission controller...

Yes, it showed up in OSE 3.2.

adellape commented on architecture/additional_concepts/admission_controllers.adoc:
```diff
@@ -32,3 +32,4 @@ ifdef::openshift-dedicated[]
 - link:../../admin_guide/osd_build_defaults_overrides.html[Configuring Global Build Defaults and Overrides]
 endif::[]
 - link:../../admin_guide/scheduler.html#controlling-pod-placement[Controlling Pod Placement]
+- link:../../admin_guide/overcommit.html[Configuring Overcommit]
```
@sosiouxme Looks like we don't include the Overcommitting topic in the Dedicated version of the Cluster Admin guide, so this new line would need to get ifdef'd appropriately (or move it up into the other dedicated ifdef). But to your point in the other comment, if you don't think it warrants a highlight here, fine to omit completely.

sosiouxme replied:
on reflection i think i will omit.

adellape commented:
@sosiouxme Thanks. 1 comment inline, other changes look good. Can you squash when ready? Then the squirrel will ship.

sosiouxme commented:
:shipit:

adellape commented:
Revhistory only really meant for OSE; picking to OSD and Online to keep files in sync (namely the dev_guide topic).

[rev_history]
|link:../admin_guide/overcommit.html[Overcommitting]
|Added the link:../admin_guide/overcommit.html#configuring-masters-for-overcommitment[Configuring Masters for Overcommitment] section about the `*ClusterResourceOverride*` admission controller.
%
|link:../dev_guide/compute_resources.html[Quotas and Limit Ranges]
|Added a section on project resource limits.
%
